Added encoding colorimetry

Bananahavoc opened the pull request:
This uses colorimetry specification present in the gstreamer caps to tell the encoder to add SEI information into the encoded bitstream. This SEI data is used by decoders.

ishmael1985 reviewed (changes requested):
You should try to follow the GStreamer coding style (spaces instead of tabs, underscores instead of camel case, K&R C). Though there may be a few parts of the codebase that may not adhere to the GStreamer coding style, which in this case you need to help point out.

ishmael1985 commented on gst-libs/mfx/gstmfxencoder.c:
```diff
@@ -406,6 +406,97 @@ gst_mfx_encoder_properties_get_default (const GstMfxEncoderClass * klass)
 
+static void
+gst_mfx_extVideoSignal_from_colorimetry(GstVideoColorimetry *colorimetry, mfxExtVideoSignalInfo *extVideo) {
+	memset(extVideo, 0, sizeof(mfxExtVideoSignalInfo));
```
Not really needed, since GObjects zero initializes all its members.

ishmael1985 commented on gst-libs/mfx/gstmfxencoder.c:
```diff
@@ -1026,6 +1120,7 @@ gst_mfx_encoder_prepare (GstMfxEncoder * encoder)
       return GST_MFX_ENCODER_STATUS_ERROR_OPERATION_FAILED;
 
   gst_mfx_encoder_set_frame_info (encoder);
+  priv->params.ExtParam = priv->extparam_internal;
```
May have to find a better place to put this statement...

Bananahavoc replied:
Any suggestions where it should go?

ishmael1985 replied:
last line of gst_mfx_encoder_set_encoding_params () seems to be the best place...

ishmael1985 commented on gst-libs/mfx/gstmfxencoder.c:
```diff
@@ -406,6 +406,97 @@ gst_mfx_encoder_properties_get_default (const GstMfxEncoderClass * klass)
+	}
+	switch (colorimetry->transfer) {
+	case GST_VIDEO_TRANSFER_BT709:
+		extVideo->TransferCharacteristics = 1;// BT709
```
MFX_TRANSFERMATRIX_BT709 ?

ishmael1985 replied:
Maybe wrong. These enumerators may have been intended for mfxExtVPPVideoSignalInfo.

Bananahavoc replied:
I believe the numbers are supposed to be from Annex E ISO*/IEC* 14496-10 http://www.staroceans.org/e-book/ISO-14496-10.pdf

ishmael1985 replied:
You might want to add bt. 2020 support..
https://sourceforge.net/p/mediainfo/feature-requests/508/

ishmael1985 commented on gst-libs/mfx/gstmfxencoder.c:
```diff
@@ -514,12 +510,12 @@ gst_mfx_encoder_set_frame_info (GstMfxEncoder * encoder)
-	priv->extparam_internal[priv->params.NumExtParam++] = (mfxExtBuffer*)&(priv->extVideoSignalInfo);
-
+    
+    gst_mfx_extsig_from_colorimetry(&GST_VIDEO_INFO_COLORIMETRY
```
80 character limit!

Bananahavoc replied:
Sorry, not used to formatting my code like this

ishmael1985 replied:
https://gstreamer.freedesktop.org/documentation/frequently-asked-questions/developing.html#what-is-the-coding-style-for-gstreamer-code

Np. It's really only to enforce consistency and readability. But code in itself looks good to me.

ishmael1985 commented:
I have tested and cleaned up your patch, seems that setting the colorimetry works only for H264 encoding quite well. Tried to set BT2020 support for HEVC Main 10 encodes but no luck, encoder crashes.
I am attaching the patch, please rebase against current master and squash all your commits with this patch.
[colorimetry.txt](https://github.com/ishmael1985/gstreamer-media-SDK/files/1800489/colorimetry.txt)

ishmael1985 commented:
Very minor revision (open bracket in next line after name of function), as attached.
[colorimetry.txt](https://github.com/ishmael1985/gstreamer-media-SDK/files/1800502/colorimetry.txt)

Commit:
Added encoding colorimetry SEI

The mfx h264 encoder now uses colorimetry data from the caps in the SEI of the encoded bitstream

ishmael1985 reviewed (commented):
LGTM

ishmael1985 commented:
@BananaHemic , https://github.com/ishmael1985/gstreamer-media-SDK/commit/3a5c60dda9f2270e628fd57e2ae1657912e27270 which includes BT. 2020 support (tested with HEVC Main 10 encoding in conjunction with qtmux), in case you didn't notice.
Thanks a lot for your contribution.
